### DISCUSSION TITLE
A more flexible and enhanced way of logging in .NET 8

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#84] [General]
# A more flexible and enhanced way of logging in .NET 8

The latest version of the .NET (version 8) has introduced a "better" way of logging. This new way of logging is more flexible and enhanced than the previous versions. It is about the LoggerMessageAttribute.

https://steven-giesel.com/blogPost/1add3827-4233-4e31-9ce9-bcc34d64e76f

**Comment by cocowalla** [comment]
> it is nicer to read with less code

The source generator version is about 25% more code than the original, not less?

**Reply by linkdotnet** [reply]
Well, technically, the source-generated code, does more.
It also does the following:
```csharp
[LoggerMessage(LogLevel.Debug, "Deleted Batch {BatchNumber}. In total {TotalDeleted} elements deleted")]
[GeneratedCode("Microsoft.Extensions.Logging.Generators", "8.0.9.3103")]
private void LogDeleteBatch(int batchNumber, int totalDeleted)
{
  if (!this.logger.IsEnabled(LogLevel.Debug))
    return;
  Repository<TEntity>.__LogDeleteBatchCallback((ILogger) this.logger, batchNumber, totalDeleted, (Exception) null);
}
```

With "less" code - I referred more to the "complexity" of the code itself. Reading `LogDeleteBatch(batch + 1, (batch + 1) * batchSize);` in contrast to `logger.LogDebug("Deleted Batch {BatchNumber}. In total {TotalDeleted} elements deleted", batch + 1, (batch + 1) * batchSize);` is way less taxing. 

I should clarify that a bit better. Point taken.

**Comment by abedkanbar** [comment]
The `LoggerMessageAttribute` has been available since .NET 6, and I have been using it since then.

**Reply by abedkanbar** [reply]
Indeed, there are constructor overloads that do not exist.

**Comment by toupswork** [comment]
Yeah, I love this feature. It doesn't cause boxing or concatenations that create extra allocations. But I still prefer to create them using extension methods in another class.

**Reply by linkdotnet** [reply]
I am currently in the mind of using partial classes just for the logging.
It removes a bit the clutter.

Extensions are The oldschool-way I also did before.
Did you try the partial class approach?

**Reply by toupswork** [reply]
I assume you mean create a partial class of all your classes and put the logging clutter there. I like having them all in one place honestly. I have one extension method logging class for the Presentation, Infrastructure, Application, and Domain layers

**Reply by linkdotnet** [reply]
Exactly- have the logging stuff in a separate file like MyService.Logger.cs

**Reply by toupswork** [reply]
Yes but some logging methods are shared, and having it in one place makes it easy to manage the "name" and error code attributes. I can see in one glance all the logs and can evaluate whether I want to adjust the attributes.
